# Data Exploration

## Data Import

In [1]:
import numpy as np
import pandas as pd


In [2]:
d_hjorring = pd.read_excel("../data/data_district_heating.xlsx", sheet_name="Hjorring")
print(d_hjorring.head())

    SerialID  building_id  zone_id     SEEB  \
0  311487854            0        0  2-1-3-0   
1  311235119            0        0  2-1-3-0   
2  311235135            0        0  2-1-3-0   
3  311235144            0        0  2-1-3-0   
4  311488415            0        0  2-1-3-0   

                                    S_text  \
0        Fjernvarme uden veksler - direkte   
1  Fjernvarme uden veksler (direkte anlæg)   
2  Fjernvarme uden veksler (direkte anlæg)   
3  Fjernvarme uden veksler (direkte anlæg)   
4        Fjernvarme uden veksler - direkte   

                                              L_text                Type  \
0  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
1  Ejendommen opvarmes med fjernvarme, og anlægge...  DirectDistrictHeat   
2  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
3  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
4  Bygningen opvarmes med fjernvarme. Anlægget er...  Direct

In [3]:
d_favrskov = pd.read_excel("../data/data_district_heating.xlsx", sheet_name="Favrskov")
print(d_favrskov.head())

    SerialID  building_id  zone_id     SEEB  \
0  311239024            0        0  2-1-3-0   
1  311239047            0        0  2-1-3-0   
2  311239049            0        0  2-1-3-0   
3  311244995            0        0  2-1-3-0   
4  311251256            0        0  2-1-3-0   

                                    S_text  \
0  Fjernvarme uden veksler (direkte anlæg)   
1  Fjernvarme uden veksler (direkte anlæg)   
2  Fjernvarme uden veksler (direkte anlæg)   
3  Fjernvarme uden veksler (direkte anlæg)   
4  Fjernvarme uden veksler (direkte anlæg)   

                                              L_text                Type  \
0  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
1  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
2  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
3  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
4  :Klynge A, B, C og D:\nBygningerne opvarmes me...  Direct

In [4]:
d_aarhus = pd.read_excel("../data/data_district_heating.xlsx", sheet_name="Aarhus")
print(d_aarhus.head())

    SerialID  building_id  zone_id     SEEB  \
0  311075771            0        0  2-1-3-0   
1  311075781            0        0  2-1-3-0   
2  311075799            0        0  2-1-3-0   
3  311075871            0        0  2-1-3-0   
4  311075881            0        0  2-1-3-0   

                                    S_text  \
0  Fjernvarme uden veksler (direkte anlæg)   
1  Fjernvarme uden veksler (direkte anlæg)   
2  Fjernvarme uden veksler (direkte anlæg)   
3  Fjernvarme uden veksler (direkte anlæg)   
4                       Direkte fjernvarme   

                                              L_text                Type  \
0  Ejendommen opvarmes med fjernvarme, og anlægge...  DirectDistrictHeat   
1  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
2  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
3                                                NaN  DirectDistrictHeat   
4                                                NaN  Direct

In [5]:
d_gt = pd.read_excel("../data/data_district_heating.xlsx", sheet_name="Ground Truth")
print(d_gt.head())

    SerialID  building_id  zone_id     SEEB  \
0  311117034            0        0  2-1-3-0   
1  311117893      2050872  2050872  2-1-3-0   
2  311119019            0        0  2-1-3-0   
3  311121828            0        0  2-1-3-0   
4  311122000            0        0  2-1-3-0   

                                              S_text  \
0  Fjernvarme med isoleret veksler (indirekte anl...   
1                            Eksisterende fjernvarme   
2  Fjernvarme med uisoleret veksler (indirekte an...   
3  Fjernvarme med isoleret veksler (indirekte anl...   
4  Fjernvarme med isoleret veksler (indirekte anl...   

                                              L_text  \
0  Bygningen opvarmes med fjernvarme. Anlægget er...   
1                                                NaN   
2  Bygningen opvarmes med fjernvarme. Anlægget er...   
3  Ejendommen opvarmes med fjernvarme fra HOFOR.\...   
4  Bygningen opvarmes med fjernvarme. Anlægget er...   

                        Type  Pieces1  Manu

## Data Size

In [6]:
print("Data Favrskov shape: ", d_favrskov.shape)
print("Data Hjorring shape: ", d_hjorring.shape)
print("Data Aarhus shape: ", d_aarhus.shape)
print("Data Ground Truth shape: ", d_gt.shape)

Data Favrskov shape:  (111, 15)
Data Hjorring shape:  (6138, 15)
Data Aarhus shape:  (29542, 15)
Data Ground Truth shape:  (2272, 42)


Let's merge the dataset on the SerialID to check whether the ground truth data are present in the rest of the samples.

In [7]:
d_f_gt = pd.merge(left=d_favrskov, right=d_gt, on="SerialID")
d_h_gt = pd.merge(left=d_hjorring, right=d_gt, on="SerialID")
d_a_gt = pd.merge(left=d_aarhus, right=d_gt, on="SerialID")
print(d_f_gt.shape)
print(d_h_gt.shape)
print(d_a_gt.shape)

(0, 56)
(0, 56)
(0, 56)


From the previous result, we can say that the ground truth is not present in data.
Thus we have 2272 labeled samples.

In [8]:
print(d_favrskov.shape[0] + d_hjorring.shape[0] + d_aarhus.shape[0])

35791


On the other hand, we have 35791 entries we want to label.

## Features of the Ground Truths

In [9]:
print(d_gt[d_gt['SerialID'] == 311795931].iloc[0, 0:31])

SerialID                                                      311795931
building_id                                                           0
zone_id                                                               0
SEEB                                                            2-1-3-0
S_text                                              Fjernvarmeforsyning
L_text                Forsyningsformen er fjernvarme som leveres af ...
Type                                          DistrictHeatWithExchanger
Pieces1                                                             2.0
Manufacturer1                                                      Reci
SubType1                                                          VT 90
HxType1                                                      Rørveksler
NominelEffectEach1                                300 kW (total 600 kW)
Year1                                                              1984
Pieces2                                                         

In [10]:
d_gt_labels = d_gt.iloc[:, 7:31]
print(d_gt_labels.columns)
print("Number of lables: ", d_gt_labels.shape[1])

Index(['Pieces1', 'Manufacturer1', 'SubType1', 'HxType1', 'NominelEffectEach1',
       'Year1', 'Pieces2', 'Manufacturer2', 'SubType2', 'HxType2',
       'NominelEffectEach2', 'Year2', 'Pieces3', 'Manufacturer3', 'SubType3',
       'HxType3', 'NominelEffectEach3', 'Year3', 'Pieces4', 'Manufacturer4',
       'SubType4', 'HxType4', 'NominelEffectEach4', 'Year4'],
      dtype='object')
Number of lables:  24


In [11]:
print(d_gt_labels.isnull().mean() * 100)

Pieces1                 0.000000
Manufacturer1           0.000000
SubType1               82.394366
HxType1                 7.262324
NominelEffectEach1     92.561620
Year1                  58.846831
Pieces2                95.774648
Manufacturer2          95.730634
SubType2               96.830986
HxType2                96.082746
NominelEffectEach2     99.691901
Year2                  96.390845
Pieces3                99.647887
Manufacturer3          99.647887
SubType3               99.779930
HxType3                99.647887
NominelEffectEach3     99.911972
Year3                  99.691901
Pieces4                99.735915
Manufacturer4          99.735915
SubType4               99.823944
HxType4                99.735915
NominelEffectEach4    100.000000
Year4                  99.823944
dtype: float64


From the previous result, we know that every entry has informations about `Pieces1` and `Manufacturer1`.
We can see that `NominalEffectEach4` is always empty, and thus we do not have information about this column.
Finally, we can see that data are very sparse, indeed very few columns are filled.

In [12]:
null_counts = d_gt_labels.isnull().sum(axis=1)
missing_values = null_counts.value_counts()

missing_index = np.array(missing_values.index)
missing_index.sort()
missing_values = np.array(missing_values[missing_index])
missing_values_cum = np.cumsum(missing_values)
percentages = missing_values_cum / missing_values_cum[-1] * 100

for i in range(missing_index.shape[0]):
    print("Number of samples: ", missing_values_cum[i], f"\tCorresponding percentage: {percentages[i]:.3f} No more than {missing_index[i]} entries missing")

Number of samples:  1 	Corresponding percentage: 0.044 No more than 3 entries missing
Number of samples:  5 	Corresponding percentage: 0.220 No more than 6 entries missing
Number of samples:  6 	Corresponding percentage: 0.264 No more than 7 entries missing
Number of samples:  7 	Corresponding percentage: 0.308 No more than 8 entries missing
Number of samples:  8 	Corresponding percentage: 0.352 No more than 10 entries missing
Number of samples:  10 	Corresponding percentage: 0.440 No more than 12 entries missing
Number of samples:  61 	Corresponding percentage: 2.685 No more than 13 entries missing
Number of samples:  65 	Corresponding percentage: 2.861 No more than 14 entries missing
Number of samples:  85 	Corresponding percentage: 3.741 No more than 16 entries missing
Number of samples:  90 	Corresponding percentage: 3.961 No more than 17 entries missing
Number of samples:  164 	Corresponding percentage: 7.218 No more than 18 entries missing
Number of samples:  396 	Corresponding p

In [13]:
print(f"Average length of S_text: {d_gt['S_text'].str.len().mean():.2f}\t standard deviation: {d_gt['S_text'].str.len().std():.2f}")
print(f"Average length of L_text: {d_gt['L_text'].str.len().mean():.2f} standard deviation: {d_gt['L_text'].str.len().std():.2f}")

Average length of S_text: 48.65	 standard deviation: 14.20
Average length of L_text: 195.77 standard deviation: 175.14


# Export Gold Label in csv

In [14]:
import sys
sys.path.append("../scripts")
from data_from_xlsx import merge_columns, to_csv

In [15]:
d_gt['text'] = d_gt['S_text'].fillna("").astype(str) + ". " + d_gt['L_text'].fillna("").astype(str)
print("Total number of sentences: ", (d_gt['text'].str.count(r'[.!?](?:\s|$)') + 1).sum())

Total number of sentences:  11017


In [16]:
print("Total number of lables: ", d_gt_labels.count().sum())

Total number of lables:  8662


In [17]:
n_sentences = (d_gt['text'].str.count(r'[.!?](?:\s|$)') + 1).sum()
n_labels = d_gt_labels.count().sum()
print("Labels on sentences ratio: ", n_labels/n_sentences, " Labels on sentences percentage: ", n_labels/n_sentences*100)

Labels on sentences ratio:  0.7862394481256241  Labels on sentences percentage:  78.62394481256241


In [18]:
for col in d_gt_labels.columns:
    print(f"{col} has the following values: {d_gt_labels[col].value_counts()}")

Pieces1 has the following values: Pieces1
1.00    2022
2.00     184
3.00      39
4.00      22
5.00       2
0.25       1
0.53       1
0.94       1
Name: count, dtype: int64
Manufacturer1 has the following values: Manufacturer1
Unknown                           1452
Reci                               212
Gemina Termix                      106
None - direct district heating      80
CTC                                 67
Sondex Teknik                       66
Reflex                              44
Danfoss Redan                       28
Danfoss                             24
Elge                                22
Kähler & Breum                      21
WPH Teknik                          21
Metro Therm                         20
Cetetherm                           20
Alfa Laval                          19
Ajva                                14
Armatec                             13
Ajva                                 6
APV                                  5
Ducon                            

In [19]:
print(d_gt.loc[195,'text'])

Fjernvarme med ISOLERET veksler (indirekte anlæg). Bygningen opvarmes med fjernvarme via varmecentral placeret i kælderen Njalsgade 12.

Varmecentral - Njalsgade 12. 
Forsyner AB Egilshus - 4 bygninger.      
Anlægget er udført med 4 stk. isoleret varmevekslere og indirekte centralvarmevand i fordelingsnettet.
Fabrikat reflex, år 2007 (1 stk.) og Fabrikat Reci, år 1989 (3 stk.).




In [20]:
from tqdm.auto import trange
import sys
sys.path.append("../scripts")
from tokenizer import Tokenizer